crates/common/tedge_config_macros/src/multi.rs

```diff
@@ -0,0 +1,209 @@
+#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(untagged)]
+pub enum MultiDto<T> {
+    Multi(::std::collections::HashMap<String, T>),
+    Single(T),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(untagged)]
+pub enum MultiReader<T> {
+    Multi {
+        map: ::std::collections::HashMap<String, T>,
+        parent: &'static str,
+    },
+    Single {
+        value: T,
+        parent: &'static str,
+    },
+}
+
+impl<T: Default> Default for MultiDto<T> {
+    fn default() -> Self {
+        Self::Single(T::default())
+    }
+}
+
+impl<T: doku::Document> doku::Document for MultiDto<T> {
+    fn ty() -> doku::Type {
+        T::ty()
+    }
+}
+
+impl<T: doku::Document> doku::Document for MultiReader<T> {
+    fn ty() -> doku::Type {
+        T::ty()
+    }
+}
+
+impl<T: Default + PartialEq> MultiDto<T> {
+    pub fn is_default(&self) -> bool {
+        *self == Self::default()
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum MultiError {
+    #[error("You are trying to access a named field ({1}) of {0}, but the fields are not named")]
+    SingleNotMulti(String, String),
+    #[error("You need a name for the field {0}")]
+    MultiNotSingle(String),
+    #[error("Key {0}.{1} not found in multi-value group")]
+    MultiKeyNotFound(String, String),
+}
```

Comment:
Thanks for adding a context to these error messages.

I would go one step further, improving the wording after [Reuben's suggestion to use the term *profile*](https://github.com/thin-edge/thin-edge.io/pull/3126#discussion_r1781215203).

```suggestion
#[derive(Debug, thiserror::Error)]
pub enum MultiError {
    #[error("Unexpected profile {1} for the non multi-profile property {0}")]
    SingleNotMulti(String, String),
    #[error("A profile is required for the multi-profile property {0}")]
    MultiNotSingle(String),
    #[error("Unknown profile {1} for the multi-profile property {0}")]
    MultiKeyNotFound(String, String),
}
```

By the way, I don't see how a user can get the `SingleNotMulti` error case. If a property is not multi, then the error will be that the key is unknown. Isn't it?